internal/gps/source_cache_bolt_encode.go

```diff
@@ -203,9 +174,13 @@ func cacheGetManifest(b *bolt.Bucket) (RootManifest, error) {
 	}
```

Comment:
There are all sorts of [`BucketStats`](https://godoc.org/github.com/boltdb/bolt#BucketStats) available via [`Stats()`](https://github.com/boltdb/bolt/blob/master/bucket.go#L398) but they have to be computed so I didn't explore that path.

For the ones using sequential integer keys (behaving like sets), we could theoretically inspect the last key to get the total items.  I don't know how the costs of jumping the cursor around, or scanning backwards, compares to re-allocating the slice.

More generally, and perhaps simply, we could just store size hints in additional top level keys.

Reply:
size hints as additional top-level keys is interesting. seems like a basically zero-cost way of doing things - that is, as long as the use pattern continues to be that these buckets tend to get dropped and wholly regenerated, rather than incrementally updated. if incremental updates become more the norm, then i'd be concerned that we're create a write amplifier. that might not be that costly, especially if the workflow is heavily read-oriented (which, of course it should be), but...well, with all these considerations, we'd need benchmarks!

probably best to leave it as-is for now, though, and put a big TODO on it noting the possible future direction to explore.